# World Bank Loan Analysis

The [data](https://www.kaggle.com/datasets/felipeea/ida-credits-grants?select=IDA_Statement_Of_Credits_and_Grants_-_Historical_Data.csv) comes from the World Bank Group. This data set contains more than a million rows and 30 columns. 
The analysis will mainly be focused on India and I will try to answer following business questions:
1. How many countries have taken loan from World Bank?
2. Top 5 countries with highest loan amount.
3. Top 5 countries with highest due amount.
4. Project first approved loan for India.
5. Top 5 project in India having highest loan.
6. Top 5 project in India having highest  due amount of loans.
7. Total number of loan projects that have fully repaid the loan.
8. Top 5 maximum approval time in project for India.
9. Top 5 minimum approval time in project for India.

### Importing necessary libraries

In [1]:
import numpy as np
import pandas as pd
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

### Reading the dataset

In [2]:
df=pd.read_csv('OneDrive/Desktop/IDA_Statement_Of_Credits_and_Grants_-_Historical_Data.csv',low_memory=False)
df.head()

End of Period Credit Number                        Region  \
0  05/31/2020 12:00:00 AM      IDA07770   LATIN AMERICA AND CARIBBEAN   
1  05/31/2020 12:00:00 AM      IDA07780                    SOUTH ASIA   
2  05/31/2020 12:00:00 AM      IDA07790                    SOUTH ASIA   
3  05/31/2020 12:00:00 AM      IDA07800  MIDDLE EAST AND NORTH AFRICA   
4  05/31/2020 12:00:00 AM      IDA07810                        AFRICA   

  Country Code      Country                                  Borrower  \
0           HN     Honduras  MINISTERIO DE HACIENDA Y CREDITO PUBLICO   
1           AF  Afghanistan                      STATE PLANNING COMM.   
2           AF  Afghanistan                      MINISTRY OF COMMERCE   
3           JO       Jordan  MINISTRY OF PLANNING & INTNL COOPERATION   
4           SD        Sudan  Ministry of Finance and National Economy   

  Credit Status  Service Charge Rate Currency of Commitment Project ID  ...  \
0  Fully Repaid                 0.75                    USD    P007353  ...   
1  Fully Repaid                 0.75                    USD    P009314  ...   
2  Fully Repaid                 0.75                    USD    P009315  ...   
3      Repaying                 0.75                    USD    P005253  ...   
4      Repaying                 0.75                    USD    P002575  ...   

  Repaid 3rd Party  Due 3rd Party  Credits Held    First Repayment Date  \
0              0.0            0.0           0.0  09/01/1988 12:00:00 AM   
1              0.0            0.0           0.0  06/15/1988 12:00:00 AM   
2              0.0            0.0           0.0  06/15/1988 12:00:00 AM   
3              0.0            0.0     3360000.0  09/01/1988 12:00:00 AM   
4              0.0            0.0    20680000.0  09/15/1988 12:00:00 AM   

      Last Repayment Date  Agreement Signing Date     Board Approval Date  \
0  03/01/2028 12:00:00 AM  04/06/1978 12:00:00 AM  03/14/1978 12:00:00 AM   
1  12/15/2027 12:00:00 AM  04/10/1978 12:00:00 AM  03/21/1978 12:00:00 AM   
2  12/15/2027 12:00:00 AM  09/27/1978 12:00:00 AM  03/21/1978 12:00:00 AM   
3  03/01/2028 12:00:00 AM  05/19/1978 12:00:00 AM  03/23/1978 12:00:00 AM   
4  03/15/2028 12:00:00 AM  06/09/1978 12:00:00 AM  03/23/1978 12:00:00 AM   

  Effective Date (Most Recent)  Closed Date (Most Recent)  \
0       08/15/1978 12:00:00 AM     12/31/1984 12:00:00 AM   
1       09/19/1978 12:00:00 AM     06/30/1983 12:00:00 AM   
2       05/15/1979 12:00:00 AM     06/30/1985 12:00:00 AM   
3       10/17/1978 12:00:00 AM     06/30/1983 12:00:00 AM   
4       02/02/1979 12:00:00 AM     06/30/1983 12:00:00 AM   

   Last Disbursement Date  
0                     NaN  
1                     NaN  
2                     NaN  
3                     NaN  
4                     NaN  

[5 rows x 30 columns]

In [3]:
df.dtypes

End of Period                    object
Credit Number                    object
Region                           object
Country Code                     object
Country                          object
Borrower                         object
Credit Status                    object
Service Charge Rate             float64
Currency of Commitment           object
Project ID                       object
Project Name                     object
Original Principal Amount       float64
Cancelled Amount                float64
Undisbursed Amount              float64
Disbursed Amount                float64
Repaid to IDA                   float64
Due to IDA                      float64
Exchange Adjustment              object
Borrower's Obligation           float64
Sold 3rd Party                  float64
Repaid 3rd Party                float64
Due 3rd Party                   float64
Credits Held                    float64
First Repayment Date             object
Last Repayment Date              object


The name of the columns are in uppercase and have spaces ans special characters, to give them a uniform look we are going to convert them all to lowercase and replace the spaces with underscores along with removing the special characters if there are any.
Also the columns are not in their respective datatypes and have to be taken care of.

### Data Preprocessing
Steps:
- converting columns to lowercase.
- casting all date columns to datetime datatype.
- replacing spaces and brackets.


In [4]:
df.columns =[x.lower() for x in df.columns]

In [5]:
df['end of period']=pd.to_datetime(df['end of period'])
df['first repayment date']=pd.to_datetime(df['first repayment date'])
df['last repayment date']=pd.to_datetime(df['last repayment date'])
df['agreement signing date']=pd.to_datetime(df['agreement signing date'])
df['board approval date']=pd.to_datetime(df['board approval date'])
df['effective date (most recent)']=pd.to_datetime(df['effective date (most recent)'])
df['closed date (most recent)']=pd.to_datetime(df['closed date (most recent)'])
df['last disbursement date']=pd.to_datetime(df['last disbursement date'])

In [6]:
df.columns = df.columns.str.replace(')','')
df.columns = df.columns.str.replace(' ','_')
df.columns = df.columns.str.replace('(','')

In [7]:
print(df.columns)
print(df.dtypes)

Index(['end_of_period', 'credit_number', 'region', 'country_code', 'country',
       'borrower', 'credit_status', 'service_charge_rate',
       'currency_of_commitment', 'project_id', 'project_name',
       'original_principal_amount', 'cancelled_amount', 'undisbursed_amount',
       'disbursed_amount', 'repaid_to_ida', 'due_to_ida',
       'exchange_adjustment', 'borrower's_obligation', 'sold_3rd_party',
       'repaid_3rd_party', 'due_3rd_party', 'credits_held',
       'first_repayment_date', 'last_repayment_date', 'agreement_signing_date',
       'board_approval_date', 'effective_date_most_recent',
       'closed_date_most_recent', 'last_disbursement_date'],
      dtype='object')
end_of_period                 datetime64[ns]
credit_number                         object
region                                object
country_code                          object
country                               object
borrower                              object
credit_status                         

Now the data types and the format is correct and we can move on to the analysis by answering questions.
<br>
### 1. How many countries have taken loan from World Bank?

In [8]:
df.country.unique().shape[0]

132

In total there are 132 countries all around the world who had taken loan.

### 2. Top 5 countries with highest loan amount

In [9]:
df[['country','original_principal_amount','cancelled_amount']].head(10)

country  original_principal_amount  cancelled_amount
0     Honduras                  5000000.0          24276.41
1  Afghanistan                 22000000.0       21912427.15
2  Afghanistan                 18000000.0       16914196.18
3       Jordan                 14000000.0              0.00
4        Sudan                 22000000.0              0.00
5        Sudan                 25000000.0        1458646.74
6         Chad                 11500000.0       11500000.00
7     Cameroon                 13000000.0              0.00
8    Indonesia                 40000000.0              0.00
9      Liberia                  6000000.0              0.00

In [10]:
df['loan'] = df['original_principal_amount'] - df['cancelled_amount']

df.groupby('country').sum().reset_index()[['country','loan']].sort_values(by='loan',ascending=False).head(5)

country          loan
54        India  5.257693e+12
6    Bangladesh  2.566757e+12
92     Pakistan  2.206612e+12
126     Vietnam  2.012682e+12
42     Ethiopia  1.879644e+12

The 5 countries with the highest loan are: India, Bangladesh, Pakistan, Vietnam and Ethiopia

### 3. Top 5 countries with highest due amount

In [11]:
df.groupby('country').sum().reset_index()[['country','due_to_ida']].sort_values(by='due_to_ida',ascending=False).head(5)

country    due_to_ida
54        India  2.930559e+12
6    Bangladesh  1.528951e+12
92     Pakistan  1.507379e+12
126     Vietnam  1.331179e+12
89      Nigeria  8.064623e+11

The top 5 countries with highest due amount are: India, Bangladesh, Pakistan, Vietnam and Nigeria

### 4. Project first approved loan in India

In [12]:
df.query('country == "India"').sort_values(by='board_approval_date')[['project_name','board_approval_date']].head()

project_name board_approval_date
311088     HIGHWAYS          1961-06-20
532076     HIGHWAYS          1961-06-20
29945      HIGHWAYS          1961-06-20
138519     HIGHWAYS          1961-06-20
303808     HIGHWAYS          1961-06-20

The first loan approved by World Bank was for the project "Highways" and was approved on 20 June 1961

### 5. Top 5 project in India having highest loan

In [13]:
india_df = df.query('country == "India"')
india_df.groupby('project_name').sum().reset_index()[['project_name','loan']].sort_values(by='loan',ascending=False).head(5)

project_name          loan
154   IN: Elementary Education (SSA II)  1.606500e+11
192       IN: PMGSY Rural Roads Project  1.100784e+11
205                         IN: SSA III  8.250840e+10
81   ELEMENTARY EDUCATION PROJECT (SSA)  5.950000e+10
389                  SOCIAL SAFETY NETS  5.950000e+10

India has taken highest loan amount for Elementary Education(SSA II) and then for Rural Roads Project.

### 6. Top 5 project in India having highest  due amount of loans.

In [14]:
india_df.groupby('project_name').sum().reset_index()[['project_name','due_to_ida']].sort_values(by='due_to_ida',ascending=False).head(5)

project_name    due_to_ida
154   IN: Elementary Education (SSA II)  1.287342e+11
192       IN: PMGSY Rural Roads Project  6.184947e+10
205                         IN: SSA III  5.594649e+10
81   ELEMENTARY EDUCATION PROJECT (SSA)  5.153341e+10
74                              DPEP II  3.705193e+10

Elementary Education and Rural Roads projects are both in the top for highest loan as well as highest due amount.

### 7. Total number of loan projects that have fully repaid the loan.

In [15]:
repaid_df = india_df.query('credit_status == "Fully Repaid"')
repaid_df['project_name'].unique().shape[0]

27

There are total 27 projects in India who have taken loan and have fully repaid it.

### 8. Top 5 maximum approval time in project for India.

In [46]:
df['processing_time'] = df['closed_date_most_recent'] - df['board_approval_date']
df.sort_values(by='processing_time',ascending=False).query('country == "India"')[['project_name','processing_time']].drop_duplicates().head(5)

project_name processing_time
24932      UPPER INDRAVATI POWE       4434 days
614699     MAHARASHTRA IRRIG. I       4186 days
705158  IN: Vocational Training       4135 days
858653             IN: RAJ WSRP       4058 days
881139     TAMIL NADU WATER SUP       3929 days

UPPER INDRAVATI POWE has the longest approval time which is approx 12 years.
### 9. Top 5 minimum approval time in project for India.

In [47]:
df.sort_values(by='processing_time',ascending=True).query('country == "India"')[['project_name','processing_time']].drop_duplicates().head(5)

project_name processing_time
323998                         IN: MIZORAM ROADS         71 days
359942                                 AP SAL II        187 days
806308  UP FISCAL REFORM & PUBLIC SECTOR RESTRCT        188 days
329648       UP FISCAL REFORM & PUBLIC SECTOR RE        188 days
53493                                 KARN SAL I        193 days

MIZORAM ROADS has the smallest approval time of 71 days.